In [1]:
import cv2 as cv
import os

archivos = [
    "images/paisaje.png",
    "images/paisaje.jpg",
    "images/paisaje.bmp",
    "images/paisaje.tiff"
]

1. Hacer un programa que cargue una imagen en distintos
formatos (por ejemplo: tiff, png, bmp, jpg).

In [2]:
for archivo in archivos:
    

    image = cv.imread(archivo)

    if image is None:
        print("No se pudo cargar:", archivo)
        continue

    print("\nArchivo:", archivo)
    print("Dimensiones:", image.shape)
    print("Tipo de datos:", image.dtype)
    print("Número de elementos:", image.size)
    print("Memoria de la matriz:", image.nbytes, "bytes")
    print("Tamaño del archivo:", os.path.getsize(archivo), "bytes")

    cv.imshow("Imagen", image)
    cv.waitKey(0)

cv.destroyAllWindows()


Archivo: images/paisaje.png
Dimensiones: (533, 800, 3)
Tipo de datos: uint8
Número de elementos: 1279200
Memoria de la matriz: 1279200 bytes
Tamaño del archivo: 876700 bytes

Archivo: images/paisaje.jpg
Dimensiones: (533, 800, 3)
Tipo de datos: uint8
Número de elementos: 1279200
Memoria de la matriz: 1279200 bytes
Tamaño del archivo: 192680 bytes

Archivo: images/paisaje.bmp
Dimensiones: (533, 800, 3)
Tipo de datos: uint8
Número de elementos: 1279200
Memoria de la matriz: 1279200 bytes
Tamaño del archivo: 1279338 bytes

Archivo: images/paisaje.tiff
Dimensiones: (533, 800, 3)
Tipo de datos: uint8
Número de elementos: 1279200
Memoria de la matriz: 1279200 bytes
Tamaño del archivo: 1279480 bytes


2. Compare visualmente las imágenes en los distintos formatos.
Indique si hay diferencias en apariencia.

Visualmente, las imágenes en formatos PNG, JPG, BMP y TIFF se observan prácticamente iguales a simple vista, ya que todas conservan las mismas dimensiones de 800 × 533 píxeles y los mismos tres canales de color. Sin embargo, internamente existen diferencias importantes en la forma de almacenamiento y compresión.

La imagen JPG ocupa mucho menos espacio en disco, con 192.680 bytes, porque utiliza compresión con pérdida. Esto puede introducir pequeñas variaciones en los píxeles, aunque normalmente no sean evidentes a simple vista. La imagen PNG ocupa 876.700 bytes y utiliza compresión sin pérdida, por lo que conserva mejor la información original. En cambio, BMP y TIFF ocupan alrededor de 1,28 MB, muy cerca del tamaño de la matriz sin comprimir, por lo que mantienen prácticamente toda la información de los píxeles.

Por tanto, aunque visualmente no se aprecien diferencias notorias entre los cuatro formatos, el tamaño de los archivos sí cambia considerablemente debido al método de compresión utilizado por cada formato.

3. Hacer una función para leer una imagen en formato RGB y la
presente en las dos escalas de grises planteadas realizando
ciclos, utilizando las siguientes ecuaciones:
- Y = 0.2989R + 0.5870G + 0.1140B
- Y = 0.333R + 0.333G + 0.333B

In [3]:
import cv2 as cv
import os


def convertir_a_gris_NTSC(image: cv.Mat) -> cv.Mat:
    if image is None:
        raise ValueError("La imagen proporcionada es None.")

    image_gris = image.copy()
    alto, ancho = image_gris.shape[:2]

    for y in range(alto):
        for x in range(ancho):
            pixel = image_gris[y, x]

            B = pixel[0]
            G = pixel[1]
            R = pixel[2]

            Y = int(0.2989 * R + 0.5870 * G + 0.1140 * B)

            image_gris[y, x] = [Y, Y, Y]

    return image_gris


def convertir_a_gris_promedio(image: cv.Mat) -> cv.Mat:
    if image is None:
        raise ValueError("La imagen proporcionada es None.")

    image_gris = image.copy()
    alto, ancho = image_gris.shape[:2]

    for y in range(alto):
        for x in range(ancho):
            pixel = image_gris[y, x]

            B = pixel[0]
            G = pixel[1]
            R = pixel[2]

            Y = int(0.333 * R + 0.333 * G + 0.333 * B)

            image_gris[y, x] = [Y, Y, Y]

    return image_gris


image = cv.imread("images/paisaje.png")

image_gris_NTSC = convertir_a_gris_NTSC(image)
image_gris_promedio = convertir_a_gris_promedio(image)

os.makedirs("resultados", exist_ok=True)

cv.imwrite(
    "resultados/paisaje_gris_ntsc.png",
    image_gris_NTSC
)

cv.imwrite(
    "resultados/paisaje_gris_promedio.png",
    image_gris_promedio
)

print("Imágenes guardadas correctamente en la carpeta resultados.")

Imágenes guardadas correctamente en la carpeta resultados.


4. Hacer un programa que lea una imagen a color y
presente la imagen en escala de grises, usando una
función para escala de grises del lenguaje de
programación utilizado.

In [4]:
def convertir_a_gris(image: cv.Mat) -> cv.Mat:
    if image is None:
        raise ValueError("La imagen proporcionada es None.")

    image_gris = image.copy()
    alto, ancho = image_gris.shape[:2]

    image_gris = cv.cvtColor(image_gris, cv.COLOR_BGR2GRAY)

    return image_gris

image = cv.imread("images/paisaje.png")
image_gris = convertir_a_gris(image)

os.makedirs("resultados", exist_ok=True)

cv.imwrite(
    "resultados/paisaje_gris.png",
    image_gris
)

True

Al comparar las imágenes obtenidas, la conversión mediante la ecuación NTSC y la función cv.cvtColor() de OpenCV producen resultados visualmente prácticamente iguales. Esto se debe a que OpenCV utiliza una ponderación de los canales RGB muy similar a la ecuación NTSC.

En cambio, en la conversión mediante el promedio de los canales RGB se observan algunas diferencias, principalmente en las zonas verdes de la imagen, como el llano y la vegetación, que aparecen más oscuras.

5. Guarde las imágenes de los ejercicios anteriores.

6. Analizar las diferencias presentadas entre las diferentes
imágenes obtenidas. Encontrar los valores promedio de la
intensidad (brillo) de las imágenes obtenidas.

In [5]:
print("----- BRILLO PROMEDIO -----")

print(
    "NTSC:",
    image_gris_NTSC.mean()
)

print(
    "Promedio RGB:",
    image_gris_promedio.mean()
)

print(
    "OpenCV:",
    image_gris.mean()
)

----- BRILLO PROMEDIO -----
NTSC: 126.16269230769231
Promedio RGB: 121.60681285178237
OpenCV: 126.69965056285179


In [6]:
ntsc_1_canal = image_gris_NTSC[:, :, 0]

diferencia = cv.absdiff(ntsc_1_canal, image_gris)

print(
    "Diferencia promedio NTSC vs OpenCV:",
    diferencia.mean()
)

print(
    "Diferencia máxima NTSC vs OpenCV:",
    diferencia.max()
)

Diferencia promedio NTSC vs OpenCV: 0.5369582551594747
Diferencia máxima NTSC vs OpenCV: 1


Los valores promedio de intensidad obtenidos fueron 126.16 para el método NTSC, 121.61 para el promedio RGB y 126.70 para la función de OpenCV. Estos resultados confirman la observación visual: la imagen obtenida mediante el promedio RGB es ligeramente más oscura, especialmente en las zonas con predominio del color verde. En cambio, las imágenes obtenidas mediante NTSC y OpenCV presentan valores de brillo muy similares, por lo que su apariencia visual es prácticamente idéntica.